### PR TITLE
feat: support new universal/deep link format

### LINKED ISSUE #17431: Support universal deep link for community link
# Feature Issue

### User Story

As a user I want to **click on a community link and go to the community** so that **I can interact with the community as soon as possible, with as little friction as possible**.

Figma design: https://www.figma.com/file/q3kZPl7OpbB7btFK7HG0Mr/Sharing-for-Mobile?type=design&node-id=690-280345&mode=design&t=tuTTqA3k1dD8yqMw-0

Crew link: https://www.notion.so/Chainmakers-b78f651f5aa04539ad4ac2a93218d1a7

parent issue: https://github.com/status-im/status-mobile/issues/15325

### PR DESCRIPTION
**Pull request by yqrashawn**
resolve #17431

### Summary
this is a starting PR to support new link format 
following issue for this feature:
- https://github.com/status-im/status-mobile/issues/17550
- https://github.com/status-im/status-mobile/issues/17551

crew: https://www.notion.so/Chainmakers-b78f651f5aa04539ad4ac2a93218d1a7?pvs=4

`join.status.im` links in docs are not changed, `status.app` is not ready for production.

scope of this PR:
1. handle `status-app://` and `https://status.app` url
open the first one in safari should trigger open the app and jump to the specific view 
open second one in app (from chat) should trigger jump to the specific view
1. when triggered before login or onboarding, app should remember the link and jump to the specific view after login
1. not working: input `https://status.app` in safari -> auto open app and jump
1. not working: generate proper profile/community url using share

detail:
1. `status-im://` -> `status-app://`
1. `https://join.status.im` -> `https://status.app`
1. jump to the right screen (profile/community overview) depends on link format
1. link to status public group chat is broken, how do we handle this? Separate issue https://github.com/status-im/status-mobile/issues/17493
https://github.com/status-im/status-mobile/blob/1e68a162bd99c063109ed06b50ac95187d79fa1a/src/status_im/ui/screens/browser/site_blocked/views.cljs#L24
1. user link jump to user profile, test link comes from [status-web](https://github.com/status-im/status-web/blob/main/packages/status-js/src/utils/create-url.test.ts) 
https://status.app/u/G10A4B0JdgwyRww90WXtnP1oNH1ZLQNM0yX0Ja9YyAMjrqSZIYINOHCbFhrnKRAcPGStPxCMJDSZlGCKzmZrJcimHY8BbcXlORrElv_BbQEegnMDPx1g9C5VVNl0fE4y#zQ3shwQPhRuDJSjVGVBnTjCdgXy5i9WQaeVPdGJD6yTarJQSj
1. community link jump to community overview 
https://status.app/c/iyKACkQKB0Rvb2RsZXMSJ0NvbG9yaW5nIHRoZSB3b3JsZCB3aXRoIGpveSDigKIg4bSXIOKAohiYohsiByMxMzFEMkYqAwEhMwM=#zQ3shYSHp7GoiXaauJMnDcjwU2yNjdzpXLosAWapPS4CFxc11
1. community channel link jump to community overview (for now, should jump to channel overview?) 
https://status.app/cc/G54AAKwObLdpiGjXnckYzRcOSq0QQAS_CURGfqVU42ceGHCObstUIknTTZDOKF3E8y2MSicncpO7fTskXnoACiPKeejvjtLTGWNxUhlT7fyQS7Jrr33UVHluxv_PLjV2ePGw5GQ33innzeK34pInIgUGs5RjdQifMVmURalxxQKwiuoY5zwIjixWWRHqjHM=#zQ3shYSHp7GoiXaauJMnDcjwU2yNjdzpXLosAWapPS4CFxc11
1. `status-app://` links should have the same behavior, `https://status.app/c/...` -> `status-app://c/...`
1. remove old unused uri routes that's not in the spec, `b/`, `p/`, `cr/`..., keep their handler for future
https://github.com/status-im/status-mobile/blob/ef3798ab9401dbeeaa5c935f8493a52c59ae10fd/src/status_im/router/core.cljs#L49

### Testing notes
Note:
for links that jump to unfetched community, the code call status-go to fetch it from mailserver, but it takes a long time for the data to show up
related desktop issue https://github.com/status-im/status-desktop/issues/12188

not sure if this happens to anyone else or this is a known issue
might be related to how long the community has been created

#### Platforms

- Android
- iOS

#### Areas that maybe impacted
anything related to `join.status.im` and `status-im:`

videos:
[jump to profile](https://github.com/status-im/status-mobile/issues/17431#issuecomment-1744995984)
[remember link and jump to community after login, community not yet loaded](https://github.com/status-im/status-mobile/issues/17431#issuecomment-1744468227)

status: ready

### CODE REVIEW

**Review by cammellos — APPROVED** [review]
Public chat will not be supported anymore, so we should not worry about them

**Comment by VolodLytvynenko** [comment]
Hi @yqrashawn Thank you for the PR. All the links, whether triggered in Safari or in the chat within Status, navigate me to the wrong page. Please take a look.

## ISSUE 1: "Login to Status App" Popup shown if  profile or community links are opened in Safari
### Steps:
1. Share profile or community link 
2. Open open current link in Safari (Profile link which is used in my case: https://status.app/u/zQ3shcRHGa8QkPEBM7R1fRmwrA5zEV94tqXMeJrCss2oQEVnm)

### Actual result:
When trying to open the link in Safari, a "Login to Status App" popup is displayed, which prevents direct navigation to the linked profile or community.

https://github.com/status-im/status-mobile/assets/52490791/d5d43167-003d-468c-9106-1839fff0666f


### Expected result:
The link navigates the user to a specific community or profile

### Device:
- iPhone 11 Pro max, IOS 16

**Comment by VolodLytvynenko** [comment]
## ISSUE 2: Links opened in Status chat do not navigate to specific profile or community

### Steps to Reproduce:
1. Share a link to a Status profile or community within a Status chat.
2. Tap the current link

### Actual Result:
When the link is tapped in the Status chat, it does not navigate to the specific profile or community 

https://github.com/status-im/status-mobile/assets/52490791/55dfcf2d-0b5a-4c56-9741-12534c8f2d23



### Expected Result:
Clicking on a link to a Status profile or community within a Status chat should navigate the user directly to the linked profile or community within the Status app.

**Comment by yqrashawn** [comment]
Hi @VolodLytvynenko, the link format that mobile really care about when parsing are

```txt
https://status.app/u/what-ever#user-id
https://status.app/c/what-ever#community-id
https://status.app/cc/what-ever#chat-id <--this is wrong
https://status.app/cc/chat-id#community-id <-- this is the right one
```

**Review by alwx — APPROVED** [review]

**Comment by qoqobolo** [comment]
Hi @yqrashawn, great work, thank you!

I'm now trying to figure out what has already been implemented for both Android and iOS.
Could you pls help me and take a look at the following questions (only for the scenario when I follow an external link atm (not inside the Status app)):

#### Q1. How do I use `status-app://` links? (both outside and inside the Status app)
> scope of this PR:
> 1. handle status-app:// and https://status.app url
open the first one in safari should trigger open the app and jump to the specific view

It's not working for me. 
Here's what I'm doing:
1. paste this link into Safari: status-app://u/Ow==#zQ3shrgSaffLisbiqj2brkDK8Le7wpbpbpzriWd8kPUmdYhs1
2. get the error `Safari cannot open the page because the address is invalid`

Tried three cases:
- The app is not installed
- The app is installed, no profile created
- Profile is created, the user is logged out

What am I doing wrong?

https://github.com/status-im/status-mobile/assets/67952253/a284b58b-9621-45f2-9309-fac92fd2bc2c

---

#### Q2. Should `https://status.app/` links be working on Android now when following them from an external app/browser?

For example, a scenario when the app is installed.
I'm following the link `https://status.app/u/Ow==#zQ3shrgSaffLisbiqj2brkDK8Le7wpbpbpzriWd8kPUmdYhs1` from the external messenger, loggin in to status.app and getting `App error: a client-side exception has occurred`.
So basically, none of the scenarios for following an external link work for Android now.
Is this expected?

https://github.com/status-im/status-mobile/assets/67952253/60d5d73a-a3e1-439b-8f7d-9ee9536e95ae

**Comment by yqrashawn** [comment]
hi @qoqobolo 

in short
- iOS should work, attached video how I test links
- Anroid is bugged

---

`status-app://` should work on iOS

`Safari cannot open the page because the address is invalid` means no app on the device can handle `status-app://` protocol
so it won't work if the app not installed, and it should work once the app is installed

https://github.com/status-im/status-mobile/assets/15090582/ca45975a-b48e-42b4-98ef-75d89bc41afb

---

`App error: a client-side exception has occurred` is an error on `status-web` page
it's caused by the `what-ever` part in `https://status.app/u/what-ever#user-id`
mobile doesn't parse the `what-ever` part, but `status-web` page do need that to work

with Status PR build, you can see in the below video, the page shows error, but the banner in safari still works (if the banner doesn't show up, try to uninstall all status apps and reinstall PR build)

[RPReplay_Final1697116183.webm](https://github.com/status-im/status-mobile/assets/15090582/529ce084-072a-4a39-ac2c-45f8d0678ba0)

---

On Android, I just tested, it's bugged, it's not working. 
Apologize, I didn't test on Android. I'll look into it.

**Comment by qoqobolo** [comment]
> status-app:// should work on iOS

It doesn't on my real device :(
Here are three cases with the app installed
1. The app is installed, no profile

https://github.com/status-im/status-mobile/assets/67952253/fe87c8af-08f5-4552-b522-23569dcada2e

2. User is logged in


https://github.com/status-im/status-mobile/assets/67952253/a2a6cc90-a2bd-4428-86d1-17c9493a0d90

3. User is logged out


https://github.com/status-im/status-mobile/assets/67952253/3bf593bc-298f-40ca-aeb4-887501ab18b7

The link seems to be correct, same as on your video


https://github.com/status-im/status-mobile/assets/67952253/b284cd66-dc6f-45df-abaa-753fb7b825f0

**Comment by qoqobolo** [comment]
> for links that jump to unfetched community, the code call status-go to fetch it from mailserver, but it takes a long time for the data to show up
related desktop issue https://github.com/status-im/status-desktop/issues/12188
not sure if this happens to anyone else or this is a known issue
might be related to how long the community has been created

Just a note: happening to me as well, but
I also noticed that sometimes if the community hangs in this state (on the `Community admins will review your request` screen), then relogin/reopening the app can trigger faster data fetching (but not always again, so might be just a coincidence).

**Comment by yqrashawn** [comment]
> ```
> https://status.app/u/what-ever#user-id
> https://status.app/c/what-ever#community-id
> https://status.app/cc/what-ever#chat-id
> ```
the last one is incorrect should be `https://status.app/cc/chat-id#community-id`
mobile app don't care about the `chat-id` part in this PR only the `community-id`

spec at https://github.com/vacp2p/rfc/pull/602/files#diff-497acddee81a09a3aa6fb5a2aa0d0585a3ac9a9628ae0e9d4855b77309a5e7cbR41

**Comment by yqrashawn** [comment]
@qoqobolo  link trigger app issue fixed, I tested on iOS simulator, iOS real device and android simulator

wrote a page with all kind of links https://yqrashawn.com/_test/status-universal-link.html

https://github.com/status-im/status-mobile/assets/15090582/3cdd4dfb-536c-4366-bc85-e7bf0e9be7be

https://github.com/status-im/status-mobile/assets/15090582/817e9451-fb40-4cf8-8afa-70547bb5b328

**Comment by qoqobolo** [comment]
@yqrashawn thanks for the page, very helpful!

#### Q3
I've got another question.  
The case is when the Status app is not installed. (UPD: same when the app is installed)
There are two `join.status.im` links:

1. For user profile https://join.status.im/u/G10A4B0JdgwyRww90WXtnP1oNH1ZLQNM0yX0Ja9YyAMjrqSZIYINOHCbFhrnKRAcPGStPxCMJDSZlGCKzmZrJcimHY8BbcXlORrElv_BbQEegnMDPx1g9C5VVNl0fE4y#zQ3shwQPhRuDJSjVGVBnTjCdgXy5i9WQaeVPdGJD6yTarJQSj

If I follow it, I'm redirected to the old website and then I get an error after pressing `Redirect me`

https://github.com/status-im/status-mobile/assets/67952253/dd2a9fb0-b25a-4e72-a020-a36802c2863e

2. For community https://join.status.im/c/iyKACkQKB0Rvb2RsZXMSJ0NvbG9yaW5nIHRoZSB3b3JsZCB3aXRoIGpveSDigKIg4bSXIOKAohiYohsiByMxMzFEMkYqAwEhMwM=#zQ3shYSHp7GoiXaauJMnDcjwU2yNjdzpXLosAWapPS4CFxc11

If I follow it, I'm immediately redirected to the app store

https://github.com/status-im/status-mobile/assets/67952253/877be624-3eb8-4704-ae6e-e17fb4fe6633

What's the difference between these two `join.status.im` links, why do I get the error in the first case, and what's the expected behavior for both?

**Comment by qoqobolo** [comment]
@yqrashawn and could you also rebase the PR pls? A critical issue in develop on iOS is fixed now

**Comment by yqrashawn** [comment]
@qoqobolo just rebased the branch

**Comment by yqrashawn** [comment]
@qoqobolo just rebased the code

**Comment by qoqobolo** [comment]
Thanks @yqrashawn, could you also take a look at [this question](https://github.com/status-im/status-mobile/pull/17480#issuecomment-1764158846) please?

**Comment by qoqobolo** [comment]
@yqrashawn 
### ISSUE 1: Links are still not working on Android if I follow links shared from a real desktop user/community from an external messenger/browser

Any scenario (when the app is installed or not, etc).
It works fine with the links from your page https://yqrashawn.com/_test/status-universal-link.html
But if I share the link from my real desktop user or test community, I get an error:

https://github.com/status-im/status-mobile/assets/67952253/4da9139d-5d84-4277-b843-6d4b9260b77a

Steps:
1. Post a real link from desktop to some external messenger
2. Follow the link

The links I'm using:
https://status.app/u/Ow==#zQ3shrgSaffLisbiqj2brkDK8Le7wpbpbpzriWd8kPUmdYhs1
https://status.app/c/Cw-ACh0KCVRlc3QgdGVzdBIFZnZkZnYYMSIHI0IwMDAxNQM=#zQ3shcpT7nnGzPvLWaYAv7DvLTiU9zNPAfZnjEKfWH3mBYowM

If I post the links in Status chats, it works fine.

**Comment by qoqobolo** [comment]
### Q 4: The app is NOT installed: when following a `https://status.app/` link, on iOS it is suggested to go to the App Store, on Android - the web opens without prompting the user to open the App Store

Do you know if this is expected?
Shouldn't we prompt the user to go to the App Store if the app is not installed on the Android device, as happens on iOS? Can this be fixed on our side?

Android:

https://github.com/status-im/status-mobile/assets/67952253/5374050b-dc24-491d-90d1-ff5304b5df22

iOS:

Screenshot 2023-10-17 at 13 05 06

### Q 5: The app IS installed: when you follow any link on iOS, the app opens immediately, and on Android the web opens with a prompt to install the Status app and go through onboarding, although the user already has the application installed and an account created.

Do you know if this is expected?
It seems on Android we should also open the app right away (as it happens on iOS), but I'm not sure.

Android:


https://github.com/status-im/status-mobile/assets/67952253/2654b9b2-baa8-4548-bf8a-134b30131272

iOS (no redirect to the web):

https://github.com/status-im/status-mobile/assets/67952253/8e7518c4-4a82-42d9-a16e-81d73651d0d8

**Review by BalogunofAfrica — APPROVED** [review]

**Comment by yqrashawn** [comment]
@qoqobolo sorry I missed Q3. I didn't notice that difference.
That the behavior defined in the backend of join.status.im page

it's at https://github.com/status-im/universal-links-handler/blob/b1bb906857c9b134e909562bb5acfaf0d0c92747/routes/index.js#L238
the old backend supports `/u` but doesn't support `/c`, so when runs into a supported URL, it will just redirect to  play store or app store

you can try this one `https://join.status.im/whatever/link/as/long/as/not/matched`, it will jump to the store as well

**Comment by yqrashawn** [comment]
@qoqobolo about Q4
Yes, it's expected. It's not in the scope of this PR. there's another issue for this https://github.com/status-im/status-mobile/issues/17550 
it's just the iOS part is done already at https://github.com/status-im/status-mobile/issues/17550#issuecomment-1757461673
this requires support from status-web, since the app is not installed yet.

**Comment by yqrashawn** [comment]
@qoqobolo for ISSUE1
the reason the page says client error or something went wrong is caused by the link
spec of links at https://github.com/vacp2p/rfc/pull/602/files#diff-497acddee81a09a3aa6fb5a2aa0d0585a3ac9a9628ae0e9d4855b77309a5e7cbR41
for user link, it's `/u/<encoded_data>#<user_chat_key>`


`https://status.app/u/Ow==#zQ3shrgSaffLisbiqj2brkDK8Le7wpbpbpzriWd8kPUmdYhs1` 
- status-web
the `Ow==` part should be valid encoded user data, `status-im/status-web`'s code will try to decode the encoded data and will show error page when the data is invalid
- status-mobile
status-mobile doesn't care about the encoded user data, we don't decode it, we use the user-chat-key and fetch user data from waku, decision made at https://www.notion.so/Usage-of-link-data-7cf6c9e2a9554761baae9a1ba84e70c2?pvs=4

once we support https://github.com/status-im/status-mobile/issues/17550, the link should trigger the app if the app is installed even the encoded part is invalid

There might be an issue if the encoded part is broken while the app is not installed. I don't think we care about it that much. I'll look into it anyway.

**Comment by qoqobolo** [comment]
Thanks for the answers @yqrashawn!

> There might be an issue if the encoded part is broken while the app is not installed. I don't think we care about it that much. I'll look into it anyway.

Okay, let me know then if/when it is or can be fixed in the current PR. 
Thanks

File: src/status_im/utils/universal_links/core.cljs [hunk 1]
```diff
@@ -81,17 +81,20 @@
...
 
-
 (rf/defn handle-navigation-to-desktop-community-from-mobile
   {:events [:handle-navigation-to-desktop-community-from-mobile]}
   [cofx deserialized-key]
-  (navigation/navigate-to cofx :community-overview deserialized-key))
+  (rf/merge
+   cofx
+   {:dispatch [:navigate-to :community-overview deserialized-key]}
+   (navigation/pop-to-root :shell-stack)))
```
Comment (yqrashawn) [inline src/status_im/utils/universal_links/core.cljs:90]: Hi @Parveshdhull, I saw you deleted the `navigate-to-community` event here https://github.com/status-im/status-mobile/pull/17586/files#diff-f3a5b4391070354e91f7f3e37afe74dee7a9653180c0dc18a2f1fbd2c85ac1dcL786

and it looks like the below pattern should be avoided
```clojure
{:dispatch [:navigate-to :community-overview deserialized-key]}
(navigation/pop-to-root :shell-stack)
```

`handle-navigation-to-desktop-community-from-mobile` is used when open universal links in app or triggered by universal link outside app

Is this pattern the right approach for this situation?

Comment (Parveshdhull) [reply src/status_im/utils/universal_links/core.cljs:90]: Hi @yqrashawn, yeah it looks good to me.
Also, I removed that event because it wasn't used anywhere, feel free to add it back.

**Comment by yqrashawn** [comment]
Hi @qoqobolo, ready to test again

the `https://status.app` auto-trigger app feature won't work
I tried to check out back to when it works at this point https://github.com/status-im/status-mobile/pull/17480#issuecomment-1763356811
and it's not working either

anyway, it's not in the scope of this PR, and it's recorded at https://github.com/status-im/status-mobile/issues/17550
so PR ready for test again

**Review by Samyoul — APPROVED** [review]
Really great job, this looks so cool.

**Comment by qoqobolo** [comment]
Tested in this PR:
- all formats (`status-im://`, `status-app://`, `https://join.status.im/`, `https://status.app/`)
- user profile, community, and community channel links
- for scenarios when: the app is not installed; the app is installed, no profile; the app is installed, the user is logged out; the app is installed, the user is logged in
- opening from an external messenger/browser 
- opening in Status chats/browser 
- on iOS and Android devices
- invalid links and QR codes scanned with a device camera (just in case)
- links from this page https://yqrashawn.com/_test/status-universal-link.html and from real desktop users and communities
- redirect to the correct web page
- redirect to the App Store (when available)
- redirect to a community you have already joined; token gated community
- links posted in different types of Status chats 

Awesome job @yqrashawn, PR can be merged 🚀
